[Lim Boon Hai] iP

boonhaii opened the pull request:
# Duii Bot
> "Your mind is for having ideas, not for holding them" - David Allen [[Source](https://dansilvestre.com/productivity-quotes)]

Duii Bot is a bot which helps you track your tasks and its:
* Text-based
* Simple to use
* Coded in Java!

To start using the bot:
1. Download the program [here](https://github.com/boonhaii/ip/releases/tag/v0.1).
2. Open up terminal, and locate the newly downloaded file on your computer.
3. Run the file by typing `java -jar ./ip.jar`.
4. Follow the instructions given by the bot and watch it keep track of your tasks for you!

You would be doing us a *great* favour by using this bot and supporting this *free* project.

Current support features:
- [x] Tracking of tasks
- [x] Tracking of timed tasks, such as Deadlines and Events
- [ ] Reminders by the bot (coming soon!) 

Oh and if you are a Java Programmer yourself, you can take this opportunity to showcase your Java skills!
Create a new Java file and write the code below before running.
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Thank you for supporting our project!

boonhaii committed:
Implemented Level 1 Functionality

boonhaii committed:
Personalize Bot

boonhaii committed:
Implement Task class and markDone functionality

boonhaii committed:
Implement Event, Deadline classes. Update Duke class

boonhaii committed:
Implement TextUiTesting files. Adjust text alignment for Duke.java

boonhaii committed:
Implement DukeException and error handling

boonhaii committed:
Implement delete, update data type, error handling

boonhaii committed:
Updated test file for Level-6

boonhaii committed:
Add Javadoc comments, error handling for invalid id for delete, markDone

boonhaii committed:
Update testfiles

boonhaii committed:
Implement save functionality for Duke

Loads/Creates .txt file upon startup.
Save tasks upon addition, deletion, or marking completion.
Update testfiles according to new functionality.

boonhaii committed:
Update Deadline and Event objects with DateTime.

Deadline and Event object previously took time as a String.

Code has now been updated such that time attribute of Event and
deadline attribute of Deadline are now LocalDateTime objects.

Duke now takes event and deadline inputs in the following manner:
* Event - "event *description* /at *DD/MM/YYYY* *24H-Time*
* Deadline - "Deadline *description* /by *DD/MM/YYYY* *24H-Time*
If time is not included in the input, a default time of 0000 is used.

Also updated input.txt and EXPECTED.TXT in text-ui-test accordingly.

boonhaii committed:
Merge branch 'branch-Level-7'

boonhaii committed:
Merge branch 'branch-Level-8'

boonhaii committed:
Implement A-MoreOOP requirements

Abstracted Duke into UI, Storage, TaskList and
Parser classes.

Created command classes which are returned
after parsing input to allow TaskList to execute
the respective method.

Created User class to implement multi-user bot usage.

Updated text-ui-test files according to updated bot
behaviour.

boonhaii committed:
Merge remote-tracking branch 'origin/master' into HEAD

boonhaii committed:
Updated text-ui-test file

Updated EXPECTED.TXT file to reflect the
correct behaviour of the current program
and with the updated inputs.

boonhaii committed:
Update .gitignore file

boonhaii committed:
Update .gitignore file

boonhaii committed:
Sort files into respective packages

boonhaii committed:
Add J-Unit Tests

Add test directory with J-Unit Tests.

Updated all other classes to allow for support
of J-Unit tests.

boonhaii committed:
Update Duke for execution as .jar application

boonhaii committed:
Add JavaDocs for all classes.

boonhaii committed:
Updated JavaDoc for DukeException.java

boonhaii committed:
Edit code to adhere to Java Coding Standards

boonhaii committed:
Implement Level-9 Functionality

Duke is now able to search for specific tasks
containing a keyword.

FindCommand class is created for the find command.

Parser has been updated to response to the find command.

UI has been updated to include methods to display information
relating to the find command.

boonhaii committed:
Edit code to adhere to Java Coding Standards

boonhaii committed:
Merge tag 'A-CodingStandard'

Implement A-CodingStandard requirements

boonhaii committed:
Merge branch 'branch-Level-9'

boonhaii committed:
Merge branch 'A-JavaDoc'

Merged with the correct version where DukeException
JavaDocs are completed.

boonhaii committed:
Integrate Gradle Into Project

Add build.gradle file for configuration of gradle for project.

To run the project with the JUnit tests, run project with
ip [test] configurations. Error would be shown if any JUnit
tests fail.

boonhaii committed:
Edit Code for Java Checkstyle

Edit code to fit requirements of Java Coding Standards.

Update build.gradle file to support checkstyle using gradle.
To run checkstyle, open gradle settings and run:
`gradle checkstyleMain checkstyleTest`
to check codestyles of main project file and test files respectively.

boonhaii committed:
Merge tag 'A-CheckStyle'

Implement A-CheckStyle requirements

LMAOboxhack commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * UI class which handles all the interactions with the user.
+ */
+public class UI {
```
```suggestion
public class Ui {
```

Should your class name be in lowercase instead? The [Naming](https://se-education.org/guides/conventions/java/intermediate.html#naming) conventions of the module coding standard state that "abbreviations and acronyms should not be uppercase when used as a part of a name".

LMAOboxhack commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,129 @@
+            file = this.getfiles();
+            FileReader reader = new FileReader(file);
+            BufferedReader buffreader = new BufferedReader(reader);
+            String text;
+            ArrayList<Task> saved = new ArrayList<>();
```
```suggestion
            String text;
            
            ArrayList<Task> saved = new ArrayList<>();
```

Might be good to add a blank line between your variable initializations and your loading logic? The [Layout](https://se-education.org/guides/conventions/java/intermediate.html#layout) section of the course coding standard states that "logical units within a block should be separated by one blank line".

LMAOboxhack commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+    /**
+     * Writes the tasks of the current session into the dynamically created file
+     * on the local filesystem.
+     *
+     * @throws DukeException If an IOException occurs during writing of file, which
+     *                       is passed on to the main() method of Duke class.
+     */
```
```suggestion
    /**
     * Writes the tasks of the current session into the dynamically created file
     * on the local filesystem.
     * 
     * @param list ...
     *
     * @throws DukeException If an IOException occurs during writing of file, which
     *                       is passed on to the main() method of Duke class.
     */
```

Would it be a good idea to add Javadoc comments for method parameters as well? Might help fellow developers understand what the inputs to your methods are. The form of Javadoc comments is specified in the [Comments](https://se-education.org/guides/conventions/java/intermediate.html#comments) section of the course coding standard.

LMAOboxhack commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,147 @@
+ */
+public class UI {
+
+    private static String logo = " ____        _ _\n"
```
```suggestion
    private static String LOGO = " ____        _ _\n"
```
Would it be good for this constant name to be capitalized? This is mentioned in the [Naming](https://se-education.org/guides/conventions/java/intermediate.html#naming) conventions of the module coding standard.

LMAOboxhack commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @return The abstract representation of the pathname of the created file
+     */
+    public File getfiles() {
```
```suggestion
    public File getFiles() {
```

Would it be good to change the method name to follow camelCase convention? This is outlined in the module's [Naming](https://se-education.org/guides/conventions/java/intermediate.html#naming) conventions.

LMAOboxhack reviewed (commented):
Looks good, just a few minor nits! 👍

boonhaii commented:
> Looks good, just a few minor nits! 👍

Thanks for the review!

boonhaii committed:
Implement GUI for Level-10.

Update Duke class with getResponse function.
Update Ui class to return String responses.
Fix Coding Standard Violations from pull requests

boonhaii committed:
Update gui directory

Update gui naming to fit Java Coding Standards

boonhaii committed:
Merge tag 'branch-Level-10'

Implement Level-10 requirements

Ongzl commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+     * Constructor of Duke instance, which initalises the other
+     * components of the project.
+     *
+     * @param user User object which encapsulates the users' name and visit history
```
think you missed a fullstop here

Ongzl commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Gets the string representation of the response by the User Interface.
+     *
+     * @param input The input string by the user
+     * @return The string to be printed out to the GUI as the response
```
Missing fullstop here too

Ongzl commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,138 @@
+     * Creates a new data directory consisting of the text file to write
+     * into if saved file does not exist.
+     *
+     * @return The abstract representation of the pathname of the created file
```
Missing fullstop here too

Ongzl commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,63 @@
+    private Task newTask;
+
+    /**
+     * Public constructor for the AddCommand.
+     * @param newTask The task to be added.
```
Think you're missing a line break here

Ongzl commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,58 @@
+    private Task removedTask;
+
+    /**
+     * Public constructor of the DeleteCommand object.
+     * @param taskid The id of the task to be removed, with respect to
```
Missing line break here too

Ongzl commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,58 @@
+    private Task doneTask;
+
+    /**
+     * Public constructor of the DoneCommand object.
+     * @param taskid The id of the task to be removed, with respect to
```
Missing line break here too

Ongzl commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,70 @@
+import duke.task.Task;
+
+/**
+ * FindCommand class which searches for tasks with matching keywords
```
Missing full stop

Ongzl commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Lists the tasks with the matching keyword associated with the FindCommand.
+     *  @param tasks The TaskList associated with the current bot.
```
Missing line break

Ongzl commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,59 @@
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Tasks which occur during a certain time
```
Missing full stop

Ongzl commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,113 @@
+     * @return The respective commands to be executed based on the input.
+     * @throws DukeException This exception is thrown when the input is invalid.
+     */
+    public static Command parse(String input) throws DukeException {
```
This method / while loop might be too long. This [coding quality website](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods)  stated the following:

>Be wary when a method is longer than the computer screen, and take corrective action when it goes
>beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

This is a little longer than 30 LOC , though I too am not sure how to condense this part of logic too haha

Ongzl reviewed (commented):
LGTM except some javadocs missing full stops and line breaks, cheers

boonhaii committed:
Implement A-Varargs requirement

Add Varargs to done and delete commands to handle
multiple ids with one command.

Abstract parse function in Parser into the various
commands to make the code cleaner.

boonhaii committed:
Merge branch 'A-Varargs'

boonhaii committed:
Update build.gradle

Update build.gradle with mainClassName to create new .jar file
consisting of GUI.

boonhaii committed:
Implement A-Assertions requirement

Add assertions to Duke, Storage, TaskList and Ui classes.

boonhaii committed:
Merge pull request #1 from boonhaii/A-Assertions

A-Assertions

boonhaii committed:
Implement A-CodeQuality requirements

Edit respective command classes to fit IS_EXIT as a constant.

Edit Storage to use Ui to display messages.
Add assert to ensure directory and file creation.

Add notification methods related to file and directory
creation in Ui.

boonhaii committed:
Merge pull request #2 from boonhaii/A-CodeQuality

Implement A-CodeQuality requirements

boonhaii committed:
Implement BCD-Extension Requirements

Add tagging functionality to Duii, which will tag tasks with a certain word with a specified tag.
Update Storage to be able to load tasks with tags.
Update FindCommand to also include tasks with the keyword as part of its tag.
Added TagCommand which encapsulates the tagging command.

Add feature in which toDos can be specified with a duration.
Update toDo class to encapsulate duration in its class.

Add help command to Duii, which introduces the bot and lists all the available commands, as well as usage examples.
Add HelpCommand class which encapsulates the command when the help command is entered.

boonhaii committed:
Create gradle.yml

boonhaii committed:
Merge pull request #4 from boonhaii/A-CI

Create gradle.yml

boonhaii committed:
Edit code to adhere to Java Coding Standards

boonhaii committed:
Merge pull request #5 from boonhaii/BCD-Extension

BCD-Extension

boonhaii committed:
Edit Storage.java to adhere to Java Coding Standards

boonhaii committed:
Merge pull request #6 from boonhaii/BCD-Extension

Edit Storage.java to adhere to Java Coding Standards

boonhaii committed:
Update Parser such that todos are not written with additional spaces.

boonhaii committed:
Update parser to solve bug of writing into save file with additional whitespace

boonhaii committed:
Update gradle.yml

boonhaii committed:
Merge pull request #10 from boonhaii/A-CI

A-CI

boonhaii committed:
Update GUI

GUI now prints startup message upon starting and exit message upon exit.

boonhaii committed:
Edit to meet Java Coding Standards

boonhaii committed:
Merge pull request #11 from boonhaii/BCD-Extension

BCD-Extension

boonhaii committed:
Update GUI

boonhaii committed:
Fix delete/done bug.

Fix bug where error is thrown in terminal when id for delete and done does not exist, for both single and multiple delete and done commands.
Fix DialogBox.java checkstyle error.

boonhaii committed:
Refactor code and add functionality

Refactor code to avoid magic string literals and apply SLAP.

Add new functionality to remove existing tags from tasks in the list.
Update help function with new functions.
Add InvalidCommand class to apply SLAP.

boonhaii committed:
Implement A-BetterGUI requirement

Update GUI with better design.
Edit code to fit Java Coding Standards.

boonhaii committed:
Update GUI

boonhaii committed:
Fix typo and update multi-word description for tagging

boonhaii committed:
Uploaded Ui.png, update application with title

boonhaii committed:
Fix bug where GUI will not display text fully

boonhaii committed:
Fix code to fit Java Coding Standard

boonhaii committed:
Update README.md

boonhaii committed:
Set theme jekyll-theme-cayman

boonhaii committed:
Update README.md

boonhaii committed:
Update README.md

boonhaii committed:
Update application with display picture credits

boonhaii committed:
Merge commit '1e05822e5f45b5b902a70b94d032bf2afec09d32'
